sdks/go/pkg/beam/core/runtime/harness/statecache/statecache.go

```diff
@@ -21,25 +21,45 @@
 package statecache
 
 import (
+	"io"
 	"sync"
 
+	"github.com/apache/beam/sdks/v2/go/pkg/beam/core/typex"
 	"github.com/apache/beam/sdks/v2/go/pkg/beam/internal/errors"
 	fnpb "github.com/apache/beam/sdks/v2/go/pkg/beam/model/fnexecution_v1"
 )
 
 type token string
 
-// ReusableInput is a resettable value, notably used to unwind iterators cheaply
-// and cache materialized side input across invocations.
+// FullValue represents the full runtime value for a data element, incl. the
+// implicit context. The result of a GBK or CoGBK is not a single FullValue.
+// The consumer is responsible for converting the values to the correct type.
+// To represent a nested KV with FullValues, assign a *FullValue to Elm/Elm2.
 //
-// Redefined from exec's input.go to avoid a cyclical dependency.
-type ReusableInput interface {
-	// Init initializes the value before use.
-	Init() error
-	// Value returns the side input value.
-	Value() interface{}
-	// Reset resets the value after use.
-	Reset() error
+// Copied from exec/fullvalue.go to avoid cyclical dependencies.
+type FullValue struct {
+	Elm  interface{} // Element or KV key.
+	Elm2 interface{} // KV value, if not invalid
+
+	Timestamp typex.EventTime
+	Windows   []typex.Window
+	Pane      typex.PaneInfo
+}
+
+// Stream is a FullValue reader. It returns io.EOF when complete, but can be
+// prematurely closed.
+//
+// Copied from exec/fullvalue.go to prevent cyclical dependencies.
+type Stream interface {
+	io.Closer
+	Read() (*FullValue, error)
```

Comment:
Copying this from exec will not work since exec.FullValue and statecache.FullValue are different types., so version of Stream won't match the exec.Stream version. Go doesn't have any co- or contra-variance in it's type system. (See https://en.wikipedia.org/wiki/Covariance_and_contravariance_(computer_science)) 

Instead this package should depend on the exec package, and if the exec package needs to interact with a cache-like object, it uses an interface on it's end instead. Afterall, that's how the SideInputAdapters, and state reader and similar work.

Reply:
To be unambiguous. 

Harness depends on exec & statecache.
statecache can depend on exec.
Exec uses a tight interface to avoid knowing internal cache details (just like how state reader operates).

Reply:
I've taken a run at fixing this and have precommits passing. I'm surprised though, the previous version with ReusableInputs did the same thing without typing issues.

Reply:
Interface types with the same method sets will be identical. However, structs with the same fields are not identical, though they are assignable to each other.

So the problem is FullValue is a struct type, which means the two different Stream types are returning different types, and that means the method isn't the same, making the Stream type not the same, making the Restream type not the same.

Subtle: yes.

Reply:
That's a good distinction to keep in mind, I appreciate the clarification.